<?xml version="1.0" ?>


<robot name="mrm_nerfed" xmlns:xacro="http://ros.org/wiki/xacro">

	<xacro:arg name="use_nominal_extrinsics" default="true" />   
	<xacro:include filename="$(find realsense2_description)/urdf/_d435i.urdf.xacro" />

	<material name="Purple_MTL">
		<color rgba="0.5294117647 0.5294117647 1.0 1.0"/> <!-- rviz green color -->
	</material>

	<material name="Plum_MTL">
		<color rgba="0.8431372549 0.6862745098 1.0 1.0"/> <!-- rviz green color -->
	</material>

	<material name="DarkGreen_MTL">
		<color rgba="0 0.3725490196 0 1.0"/> <!-- rviz green color -->
	</material>

	<material name="DeepSkyBlue_MTL">
		<color rgba="0 0.3725490196 0.5294117647 1.0"/> <!-- rviz green color -->
	</material>

	<material name="MediumSpringGreen_MTL">
		<color rgba="0 1.0 0.6862745098 1.0"/> <!-- rviz green color -->
	</material>	

	<material name="LightPink_MTL">
		<color rgba="0.8431372549 0.5294117647 0.5294117647 1.0"/> <!-- rviz green color -->
	</material>

	<link name="world">
	</link>

	<joint name="fixed_joint" type="fixed">
		<origin xyz = "0.0 0.0 0.0" rpy = "0 0 0"/>
		<parent link="world"/> <!-- <parent link="${parent}"/> -->
		<child link="base_link"/>
	</joint>

	<link name="base_link">
	</link>

	<joint name="arm_base_joint" type="fixed">
		<origin xyz = "0.0 0.0 0.06" rpy = "0 0 0"/>
		<parent link="base_link"/> <!-- <parent link="${parent}"/> -->
		<child link="arm_base_link"/>
	</joint>

	<!-- for Base -->

	<link name="arm_base_link">
		<inertial>
			<mass value = "0.286"/> <!-- kg -->
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
		 	<inertia ixx="0.0001525" ixy="0.000000" ixz="0.000000"
					 iyx="0.000000" iyy="0.0001525" iyz="0.000000"
					 izx="0.000000" izy="0.000000" izz="0.000000" />
			</inertial>
		<collision>
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
			<geometry>
				<box size="0.1 0.1 0.1"/>
			</geometry>
		</collision>
		<visual>
			<origin xyz = "-0.115 0.115 0.110" rpy = "3.1416 0 0"/> 
			<geometry>
				<mesh filename="package://mrm_nerfed/meshes/base0.5.STL" scale="0.001 0.001 0.001"/>
			</geometry>
			<material name="Purple_MTL"/> <!-- rviz will use this color -->			
		</visual>
	</link>

	<!-- for Base to Link 1 -->

	<joint name="link_1_joint" type="revolute">
		<axis xyz="0 0 1"/>
		<limit effort="1000.0" lower="-2.0944" upper="2.0944" velocity="0.5"/>
		<origin xyz = "0 0 0.08" rpy = "0 0 0"/> 
		<parent link="arm_base_link"/>
		<child link="link_1"/>
	</joint>

	<transmission name="link_1_trans">
		<type>transmission_interface/SimpleTransmission</type>
		<joint name="link_1_joint">
			<hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
		</joint>
		<actuator name="link_1_motor">
			<hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
			<mechanicalReduction>1</mechanicalReduction>
		</actuator>
	</transmission>

	<!-- for Link 1 -->

	<link name="link_1">
		<inertial>
			<mass value = "0.285"/>
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
			<inertia ixx="0.000536" ixy="0.000000" ixz="0.000000" 
					 iyx="0.000000" iyy="0.000536" iyz="0.000000"
					 izx="0.000000" izy="0.000000" izz="0.000000" />
		</inertial>

		<collision>
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
			<geometry>
				<box size="0.1 0.1 0.1"/>
			</geometry>
		</collision>
		
		<visual>
			<origin xyz = "-0.069 0.069 0.13" rpy = "3.14 0 0"/>
			<geometry>
				<mesh filename="package://mrm_nerfed/meshes/platform0.5_newest.STL" scale="0.001 0.001 0.001"/>
			</geometry>
			<material name="Plum_MTL"/> <!-- rviz will use this color -->
		</visual>
	</link>

	<!-- for Link 1 to Link 2 -->

	<joint name="link_2_joint" type="revolute">
		<axis xyz="0 0 1"/>
		<limit effort="1000.0" lower="0.0000" upper="3.1416" velocity="0.5"/>
		<origin xyz="0 0 0.1025" rpy="1.5708 0 1.5708"/>
		<parent link="link_1"/>
		<child link="link_2"/>
	</joint>

	<transmission name="link_2_trans">
		<type>transmission_interface/SimpleTransmission</type>
		<joint name="link_2_joint">
			<hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
		</joint>
		<actuator name="link_2_motor">
			<hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
			<mechanicalReduction>1</mechanicalReduction>
		</actuator>
	</transmission>

	<!-- for Link 2 -->

	<link name="link_2">
		<inertial>
			<mass value = "0.134"/>
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
			<inertia ixx="0.000000" ixy="0.000000" ixz="0.000000" 
					 iyx="0.000000" iyy="0.00000129" iyz="0.000000"
					 izx="0.000000" izy="0.000000" izz="0.00000129" />
		</inertial>
		<collision>
			<origin xyz="0 0 0" rpy="0 0 0"/>
			<geometry>
				<box size="0.1 0.1 0.1"/>
			</geometry>
		</collision>
		
		<visual>
			<origin xyz="0.32 0.025 -0.069" rpy="0 0 -3.1416"/>
			<geometry>
				<mesh filename="package://mrm_nerfed/meshes/conneting link0.5 thin.STL" scale="0.001 0.001 0.001" />
			</geometry>
			<material name="Purple_MTL"/> <!-- rviz will use this color -->
		</visual>
	</link>


	<!-- for Link 2 to Link 3 -->

	<joint name="link_3_joint" type="revolute">
		<axis xyz="0 0 1"/>
		<limit effort="1000.0" lower="-2.96706" upper="0" velocity="0.5"/>
		<origin xyz = "0.30 0 0" rpy = "0 0 0"/> <!-- <origin xyz = "0.161265 0.027938 0.118295" rpy = "0 0 0"/> -->
		<parent link="link_2"/>
		<child link="link_3"/>
	</joint>

	<transmission name="link_3_trans">
		<type>transmission_interface/SimpleTransmission</type>
		<joint name="link_3_joint">
			<hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
		</joint>
		<actuator name="link_3_motor">
			<hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
			<mechanicalReduction>1</mechanicalReduction>
		</actuator>
	</transmission>

	<!-- for Link 3 -->

	<link name="link_3">
		<inertial>
			<mass value = "0.127"/>
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
			<inertia ixx="0.000000" ixy="0.000000" ixz="0.000000" 
					 iyx="0.000000" iyy="0.00000129" iyz="0.000000"
					 izx="0.000000" izy="0.000000" izz="0.00000129" />
		</inertial>

		<collision>
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
			<geometry>
				<box size="0.1 0.1 0.1"/>
			</geometry>
		</collision>
		
		<visual>
			<origin xyz = "0.311 -0.079 0.0425" rpy = "0 3.14 -0.185"/>
			<geometry>
				<mesh filename="package://mrm_nerfed/meshes/elbow ext0.5 thin.STL" scale="0.001 0.001 0.001" />
			</geometry>
			<material name="Plum_MTL"/> <!-- rviz will use this color -->
		</visual>
	</link>



	<!-- for Link 3 to 4 -->

	<joint name="link_4_joint" type="revolute">
		<axis xyz="0 0 1"/>
		<limit effort="1000.0" lower="-2.26893" upper="1.5708" velocity="0.5"/>
		<origin xyz = "0.305 0 0" rpy = "0 0 0"/> <!-- <origin xyz = "0.149993 0.027599 0.001399" rpy = "0 0 0"> -->
		<parent link="link_3"/>
		<child link="link_4"/>
	</joint>

	<transmission name="link_4_trans">
		<type>transmission_interface/SimpleTransmission</type>
		<joint name="link_4_joint">
			<hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
		</joint>
		<actuator name="link_4_motor">
			<hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
			<mechanicalReduction>1</mechanicalReduction>
		</actuator>
	</transmission>


	<!-- for Link 4 -->

	<link name="link_4">
		<inertial>
			<mass value = "0.087"/>
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
			<inertia ixx="0.000000" ixy="0.000000" ixz="0.000000" 
					 iyx="0.000000" iyy="0.00000129" iyz="0.000000"
					 izx="0.000000" izy="0.000000" izz="0.00000129" />
		</inertial>

		<collision>
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
			<geometry>
				<box size="0.1 0.1 0.1"/>
			</geometry>
		</collision>
		
		<visual>
			<origin xyz = "0.071 0.03 -0.0675" rpy = "0 0 3.1416"/>
			<geometry>
				<mesh filename="package://mrm_nerfed/meshes/heading0.5 new.STL" scale="0.001 0.001 0.001" />
			</geometry>
			<material name="Purple_MTL"/> <!-- rviz will use this color -->
		</visual>
	</link>

	<!-- for Base Link to Tripod -->

	<joint name="tripod_joint" type="fixed">
		<origin xyz = "0.0 0.185 0.0" rpy = "0 0 0"/>
		<parent link="base_link"/> <!-- <parent link="${parent}"/> -->
		<child link="tripod_link"/>
	</joint>

	<!-- for Tripod -->

	<link name="tripod_link">
		<inertial>
			<mass value = "0.500"/> <!-- kg -->
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
		 	<inertia ixx="0.0001525" ixy="0.000000" ixz="0.000000"
					 iyx="0.000000" iyy="0.0001525" iyz="0.000000"
					 izx="0.000000" izy="0.000000" izz="0.000000" />
			</inertial>
		<collision>
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
			<geometry>
				<box size="0.1 0.1 0.1"/>
			</geometry>
		</collision>
		<visual>
			<origin xyz = "0 0 0" rpy = "0 0 0"/> 
			<geometry>
				<mesh filename="package://mrm_nerfed/meshes/tripod4.stl" scale="0.001 0.001 0.001"/>
			</geometry>
			<material name="MediumSpringGreen_MTL"/> <!-- rviz will use this color -->
		</visual>
	</link>

	<!-- for Base Link to Slab -->

	<joint name="slab_joint" type="fixed">
		<origin xyz = "0.0 0.0 0.0" rpy = "0 0 0"/>
		<parent link="base_link"/> <!-- <parent link="${parent}"/> -->
		<child link="slab_link"/>
	</joint>

	<!-- for Tripod -->

	<link name="slab_link">
		<inertial>
			<mass value = "0.500"/> <!-- kg -->
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
		 	<inertia ixx="0.0001525" ixy="0.000000" ixz="0.000000"
					 iyx="0.000000" iyy="0.0001525" iyz="0.000000"
					 izx="0.000000" izy="0.000000" izz="0.000000" />
			</inertial>
		<collision>
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
			<geometry>
				<box size="0.1 0.1 0.1"/>
			</geometry>
		</collision>
		<visual>
			<origin xyz = "0 0 0" rpy = "0 0 0"/> 
			<geometry>
				<mesh filename="package://mrm_nerfed/meshes/Stand.stl" scale="0.001 0.001 0.001"/>
			</geometry>
			<material name="MediumSpringGreen_MTL"/> <!-- rviz will use this color -->
		</visual>
	</link>

	<!-- for End-effector to Laser -->

	<joint name="laser_joint" type="fixed">
		<origin xyz = "0.0 0.0 0.0" rpy = "0 0 0"/>
		<parent link="link_4"/> <!-- <parent link="${parent}"/> -->
		<child link="laser_link"/>
	</joint>

	<!-- for Tripod -->

	<link name="laser_link">
		<inertial>
			<mass value = "0.001"/> <!-- kg -->
			<origin xyz = "0 0 0" rpy = "0 0 0"/>
		 	<inertia ixx="0.0000001" ixy="0.000000" ixz="0.000000"
					 iyx="0.000000" iyy="0.0000001" iyz="0.000000"
					 izx="0.000000" izy="0.000000" izz="0.000000" />
		</inertial>
		<visual>
			<origin xyz = "0 0 0.0" rpy = "0 1.5708 0"/> 
			<geometry>
				<mesh filename="package://mrm_nerfed/meshes/laserline.stl" scale="0.001 0.001 0.001"/>
			</geometry>
			<material name="LightPink_MTL"/> <!-- rviz will use this color -->
		</visual>
	</link>

	<xacro:sensor_d435i name="realsense" parent="tripod_link" use_nominal_extrinsics="$(arg use_nominal_extrinsics)" topics_ns = "realsense" >
		<origin xyz="0 0 0.185" rpy="0 0 1.5708"/>
	</xacro:sensor_d435i>

</robot>